### PR TITLE
Direct Memory Access

### PR DESCRIPTION
**Pull request by vdwjeremy**
While troubleshooting dynarmic high cpu usage, I noticed it goes through a software page table for all memory access.
The reason for it, I'm guessing, is to take the most out the host OS allowed allocation space while still mapping the full 64 bit addressing space.

However this comes at a cost : any memory access requires
- 1 store/load for the page indirection
- 1 store/load for the target address
- 1 compare and jump for the fallback
In addition to it the page size in yuzu is 4K, which makes a large page table, that can accelerate CPU cache fill.

The switch addressing space being 512GB, I was wondering if it would be possible to map directly the guest virtual memory to the host virtual memory. That would reduce memory accesses to a single load/store with immediate offset, it would also move the page table logic to the hardware and solve the unaligned reads/writes.
The feature is not available yet in dynarmic A64, but it could be adapted from the A32 version.

The only diff in the PR for now is a test that Windows can allocate 512GB of contiguous virtual adress space.
I'm opening this PR to start the discussion and get your feedback, in case this was already considered and discarded, or if other part of the project could be incompatible.

### CODE REVIEW

**Comment by MerryMage** [comment]
Hi! Thanks for attempting to solve this issue.

1. The major road-block to this approach for which you will have to find a solution for is as follows:
    The address space allocation granularity on Windows is 64 KiB.
2. You are incorrect when you say a fastmem approach will only require one load/store instruction. You will also have to verify whether the load/store is within the valid address space, which means this requires a minimum of 3 instructions.
    Otherwise you're providing a trivial read/write gadget which a malicious guest applicaiton can exploit.

I would be interested in hearing your ideas on the Windows address space granuarlity issue.

File: src/common/virtual_buffer.cpp [hunk 1]
```diff
@@ -22,9 +22,20 @@
 
 namespace Common {
 
+const VAddr address_space_size{1ULL << 39};
+void* vmem{nullptr};
```
Comment (lioncash) [inline src/common/virtual_buffer.cpp:26]: Please don't introduce mutable file-scope variables into the common library. We almost have all global variables removed from the codebase and adding more of them hinders this goal.

**Comment by MerryMage** [comment]
For documentation purposes I will try to clarify the constraints here.

1. One needs to allocate an appopriately sized block of shared memory with `CreateFileMappingA`. Shared memory is required for memory mirroring.
2. For each address space, one needs to allocate the 512GB region with `VirtualAlloc2`. I would advise using `MEM_RESERVE_PLACEHOLDER` for concurrency reasons.
    VirtualAlloc has a reservation granularity of 64KiB and a commit granularity of 4KiB.
3. For each region you wish to map, you must partition the previously reserved address space with `VirtualFree(MEM_RESERVE_PLACEHOLDER)`. Reading the MSDN documentation will clarify why this is necessary. You would also map in the previously allocated shared memory with `MapViewOfFile3`.
    MapViewOfFile has a granularity of 64KiB.
    I note that `VirtualProtect` has a granularity of 4KiB, which can be useful emulate a partial allocation. This gets tricky to handle, however.
4. You can unmap mapped memory with `UnmapViewOfFile2`.

You may be tempted to use `VirtualQuery`, but I would advise manually tracking state because calling into the kernel often isn't good for performance.

File: src/common/virtual_buffer.cpp (hunk 1 above)
Comment (vdwjeremy) [reply src/common/virtual_buffer.cpp:26]: alright, I'll keep that in mind when going forward

**Comment by vdwjeremy** [comment]
Thanks for your feedback

Good point for the virtual address higher range, a malicious/bad program could overflow it. We could use a bitfield to limit the address space.
Concerning the pages alignment I'm not sure to see where the problem will be
- The memory sections defined in [address_space_info.cpp#L31](https://github.com/yuzu-emu/yuzu/blob/88141bb2d44f26d1015be2cd832284bf322bc3b2/src/core/hle/kernel/memory/address_space_info.cpp#L31) are also aligned with 64K
- Mutexes are aligned on 4 bytes [mutex.cpp#L69](https://github.com/yuzu-emu/yuzu/blob/5ddcc1b2f582b10ac52ba9d329bfb4e49360c738/src/core/hle/kernel/mutex.cpp#L69), so the behaviour is the same
Does it have something to do with the way RasterizerCachedMemory is mapped ? [#L491](https://github.com/yuzu-emu/yuzu/blob/b8f5c71f2d7f819821acf036175cce65ab1ae12c/src/core/memory.cpp#L491)

**Comment by amayra** [comment]
how about linux ?

**Comment by MerryMage** [comment]
> Concerning the pages alignment I'm not sure to see where the problem will be

Switch pages are 4KiB in size and need to be placed on 4KiB boundaries. If you can only place a page at 64KiB boundaries, this poses issues.

**Comment by vdwjeremy** [comment]
Implementation started in #MerryMage/dynarmic/pull/528

**Comment by bunnei** [comment]
@vdwjeremy Do you plan to revisit this change?

**Comment by v1993** [comment]
It probably should be noted that https://github.com/MerryMage/dynarmic/tree/fm incorporating above PR exists.

**Comment by vdwjeremy** [comment]
hi @bunnei,
sorry I don't think I will be able to invest much time in this feature in the coming months,
if you are willing to give it a go, indeed https://github.com/MerryMage/dynarmic/tree/fm implements fastmem which can be used to create a flat addressing space.
I would advise benchmarking the concept before making architectural changes as perfomance gains seems to vary a lot depending on the usage.

**Comment by degasus** [comment]
Hi, I'm sorry for missing this PR. We have picked your dynarmic patch and it is in their master now. Thanks a lot.
However this one here sounds like an early proof of concept (just like my first linux one). So I think it is fine to close those one here and point to #6422 for the full implementation.

Thanks a lot for your work on fastmem. Now half a year later, you have provided the first patches to get a huge speed up.

**Pull request closed without merging**